PiT 24.9: Cannot run app in prod mode local. No instructions

Cannot run app in prod mode local, and there are no instructions in README.

Not sure how it should be performed, by I tried this:

```
mvn clean package -Pproduction -DskipTests
docker run --name postgres-dev   \
  -e POSTGRES_USER=postgres   -e POSTGRES_PASSWORD=postgres   -e POSTGRES_DB=demo   \
   -p 5432:5432   -d postgres:15

java -jar target/*.jar   \
  --spring.datasource.url=jdbc:postgresql://localhost:5432/demo   \
  --spring.datasource.username=postgres   \
  --spring.datasource.password=postgres
```

And I get this error:

```
Caused by: java.lang.IllegalStateException: Development profile is active but Vaadin is running in production mode. This indicates a configuration error - development profile should not be used in production.
	at com.example.application.security.dev.DevSecurityConfig.lambda$developmentLoginConfigurer$df50fa7c$1(DevSecurityConfig.java:73) ~[!/:1.0-SNAPSHOT]
	at com.vaadin.flow.server.VaadinService.lambda$init$0(VaadinService.java:284) ~[flow-server-24.8.1.jar!/:24.8.1]
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133) ~[na:na]
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1939) ~[na:na]
	at java.base/java.util.stream.Streams$ConcatSpliterator.forEachRemaining(Streams.java:735) ~[na:na]
	at java.base/java.util.stream.Streams$ConcatSpliterator.forEachRemaining(Streams.java:735) ~[na:na]
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:762) ~[na:na]
	at com.vaadin.flow.server.VaadinService.lambda$init$1(VaadinService.java:284) ~[flow-server-24.8.1.jar!/:24.8.1]
	at com.vaadin.flow.server.VaadinService.runWithServiceContext(VaadinService.java:2664) ~[flow-server-24.8.1.jar!/:24.8.1]
	at com.vaadin.flow.server.VaadinService.init(VaadinService.java:282) ~[flow-server-24.8.1.jar!/:24.8.1]
	at com.vaadin.flow.spring.SpringVaadinServletService.init(SpringVaadinServletService.java:172) ~[vaadin-spring-24.8.1.jar!/:na]
	at com.vaadin.flow.spring.SpringServlet.createServletService(SpringServlet.java:115) ~[vaadin-spring-24.8.1.jar!/:na]
	at com.vaadin.flow.server.VaadinServlet.createServletService(VaadinServlet.java:336) ~[flow-server-24.8.1.jar!/:24.8.1]
	at com.vaadin.flow.server.VaadinServlet.init(VaadinServlet.java:132) ~[flow-server-24.8.1.jar!/:24.8.1]
	at org.apache.catalina.core.StandardWrapper.initServlet(StandardWrapper.java:843) ~[tomcat-embed-core-10.1.42.jar!/:na]
```

--- a/walking-skeleton/Dockerfile
+++ b/walking-skeleton/Dockerfile
@@ -1,4 +1,4 @@
 FROM eclipse-temurin:21-jre
 COPY target/*.jar app.jar
 EXPOSE 8080
-ENTRYPOINT ["java", "-jar", "/app.jar"]
+ENTRYPOINT ["java", "-jar", "/app.jar", "--spring.profiles.active=prod"]

--- a/walking-skeleton/README.md
+++ b/walking-skeleton/README.md
@@ -15,6 +15,12 @@ To build the application in production mode, run:
 ./mvnw -Pproduction package
 ```
 
+To also build a Docker image, continue by running:
+
+```bash
+docker build -t my-application:latest .
+```
+
 ## Getting Started
 
 The [Getting Started](https://vaadin.com/docs/latest/getting-started) guide will quickly familiarize you with your new

--- a/walking-skeleton/src/main/java/com/example/application/security/controlcenter/ControlCenterSecurityConfig.java
+++ b/walking-skeleton/src/main/java/com/example/application/security/controlcenter/ControlCenterSecurityConfig.java
@@ -5,6 +5,7 @@ import org.springframework.boot.autoconfigure.condition.ConditionalOnCloudPlatfo
 import org.springframework.boot.cloud.CloudPlatform;
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.Profile;
 import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
 import org.springframework.security.oauth2.client.oidc.userinfo.OidcUserRequest;
 import org.springframework.security.oauth2.client.oidc.userinfo.OidcUserService;
@@ -23,7 +24,8 @@ import org.springframework.util.StringUtils;
  * </p>
  *
  * <p>
- * The configuration is conditionally activated when the application is deployed on the Kubernetes platform.
+ * The configuration is conditionally activated when the application is deployed on the Kubernetes platform
+ * <strong>and</strong> the {@code prod} Spring profile is active.
  * </p>
  *
  * <p>
@@ -36,7 +38,8 @@ import org.springframework.util.StringUtils;
 @EnableWebSecurity
 @Configuration
 @ConditionalOnCloudPlatform(CloudPlatform.KUBERNETES)
-public class ControlCenterSecurityConfig extends IdentityManagementConfiguration {
+@Profile("prod")
+class ControlCenterSecurityConfig extends IdentityManagementConfiguration {
 
     /**
      * Creates and configures an OIDC user service with custom user mapping for the application.

--- a/walking-skeleton/src/main/java/com/example/application/security/dev/DevSecurityConfig.java
+++ b/walking-skeleton/src/main/java/com/example/application/security/dev/DevSecurityConfig.java
@@ -1,16 +1,17 @@
 package com.example.application.security.dev;
 
-import com.example.application.security.controlcenter.ControlCenterSecurityConfig;
 import com.vaadin.flow.router.RouteConfiguration;
 import com.vaadin.flow.server.VaadinServiceInitListener;
 import com.vaadin.flow.spring.security.VaadinAwareSecurityContextHolderStrategyConfiguration;
 import com.vaadin.flow.spring.security.VaadinSecurityConfigurer;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
-import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.boot.cloud.CloudPlatform;
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
 import org.springframework.context.annotation.Import;
+import org.springframework.context.annotation.Profile;
+import org.springframework.core.env.Environment;
 import org.springframework.security.config.annotation.web.builders.HttpSecurity;
 import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
 import org.springframework.security.core.userdetails.UserDetailsService;
@@ -27,7 +28,7 @@ import org.springframework.security.web.SecurityFilterChain;
  * </ul>
  * </p>
  * <p>
- * This configuration is automatically activated when {@link ControlCenterSecurityConfig} is not active. It should
+ * This configuration is automatically activated when the {@code prod} Spring profile is not active. It should
  * <strong>not</strong> be used in production environments, as it uses hardcoded credentials and simplified security
  * settings.
  * </p>
@@ -46,13 +47,20 @@ import org.springframework.security.web.SecurityFilterChain;
 @EnableWebSecurity
 @Configuration
 @Import({ VaadinAwareSecurityContextHolderStrategyConfiguration.class })
-@ConditionalOnMissingBean(ControlCenterSecurityConfig.class)
+@Profile("!prod")
 class DevSecurityConfig {
 
     private static final Logger log = LoggerFactory.getLogger(DevSecurityConfig.class);
 
-    DevSecurityConfig() {
-        log.warn("Using DEVELOPMENT security configuration. This should not be used in production environments!");
+    DevSecurityConfig(Environment environment) {
+        if (!isRunningLocally(environment)) {
+            log.error("Development security config attempted in non-local environment");
+            throw new IllegalStateException("Development security can only be used when running locally");
+        }
+        log.warn("╔═════════════════════════════════════════════════════════════╗");
+        log.warn("║                     DEVELOPMENT SECURITY                    ║");
+        log.warn("║ This should not be used in production environments.         ║");
+        log.warn("╚═════════════════════════════════════════════════════════════╝");
     }
 
     @Bean
@@ -70,11 +78,23 @@ class DevSecurityConfig {
     VaadinServiceInitListener developmentLoginConfigurer() {
         return (serviceInitEvent) -> {
             if (serviceInitEvent.getSource().getDeploymentConfiguration().isProductionMode()) {
-                throw new IllegalStateException(
-                        "Development profile is active but Vaadin is running in production mode. This indicates a configuration error - development profile should not be used in production.");
+                log.warn("╔════════════════════════════════════════════════════════════════════════════════════════╗");
+                log.warn("║ DEVELOPMENT SECURITY is ACTIVE but Vaadin is running in PRODUCTION mode.               ║");
+                log.warn("║ If you are testing production mode on your local machine, this is fine.                ║");
+                log.warn("║ If you are seeing this in production, you should check your application configuration! ║");
+                log.warn("╚════════════════════════════════════════════════════════════════════════════════════════╝");
             }
             var routeConfiguration = RouteConfiguration.forApplicationScope();
             routeConfiguration.setRoute(DevLoginView.LOGIN_PATH, DevLoginView.class);
         };
+    }
+
+    private boolean isRunningLocally(Environment environment) {
+        boolean hasUserHome = System.getProperty("user.home") != null;
+        CloudPlatform activePlatform = CloudPlatform.getActive(environment);
+
+        log.info("Local environment check - User home: {}, Cloud platform: {}", hasUserHome, activePlatform);
+
+        return hasUserHome && activePlatform == null;
     }
 }

